jax/_src/lru_cache.py

```diff
@@ -0,0 +1,172 @@
+# Copyright 2024 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import heapq
+import logging
+import os
+import pathlib
+
+from jax._src import config
+from jax._src.compilation_cache_interface import CacheInterface
+
+
+try:
+  import filelock
+except ImportError:
+  filelock = None
+
+
+logger = logging.getLogger(__name__)
+
+
+class Impl:
+  """Implementation of an LRU cache, including cache reading, writing and
+  eviction based on the LRU policy.
+
+  Args:
+    path: The path to the cache directory.
+    max_size: The maximum size of the cache in bytes. A special value of
+      ``-1`` indicates no limit, allowing the cache size to grow indefinitely.
+    lock_timeout_secs: (optional) The timeout for acquiring a file lock.
+  """
+
+  def __init__(self, path: str, *, max_size: int, lock_timeout_secs: float | None = 10):
+    self.path = pathlib.Path(path)
+    self.path.mkdir(parents=True, exist_ok=True)
+
+    self.should_evict = max_size != -1
+
+    if self.should_evict:
+      if filelock is None:
+        raise RuntimeError("Please install filelock package to set `jax_compilation_cache_max_size`")
+
+      self.max_size = max_size
+      self.lock_timeout_secs = lock_timeout_secs
+
+      self.lock_file = self.path / ".lock_path"
+      self.lock = filelock.FileLock(self.lock_file)
+
+  def get(self, key: str) -> bytes | None:
+    """Retrieves the cached value for the given key.
+
+    Args:
+      key: String. The key for which the cache value is retrieved.
```

Comment:
It's unnecessary to document the types in the docstring, since you already have them in type annotations. Same for `put`.

Reply:
Done